### PR TITLE
Implement -Zlink-directives=yes/no

### PR DESCRIPTION
**Pull request by jsgf**
`-Zlink-directives=no` will ignored `#[link]` directives while compiling a crate, so nothing is emitted into the crate's metadata.  The assumption is that the build system already knows about the crate's native dependencies and can provide them at link time without these directives.

This is another way to address issue # #70093, which is currently addressed by `-Zlink-native-libraries` (implemented in #70095). The latter is implemented at link time, which has the effect of ignoring `#[link]` in *every* crate. This makes it a very large hammer as it requires all native dependencies to be known to the build system to be at all usable, including those in sysroot libraries. I think this means its effectively unused, and definitely under-used.

Being able to control this on a crate-by-crate basis should make it much easier to apply when needed.

I'm not sure if we need both mechanisms, but we can decide that later.

cc @pcwalton @cramertj

### CODE REVIEW

**Comment by jsgf** [comment]
Also cc @joshtriplett - I think I prefer this approach compared to raw_dylib since it can be controlled by the build system and doesn't need source changes.

**Comment by lcnr** [comment]
r? compiler

**Comment by joshtriplett** [comment]
I'm not sure if this is the approach we want long-term, but zero objection to it as a -Z flag for experimentation.

File: compiler/rustc_session/src/options.rs [hunk 1]
```diff
@@ -1401,6 +1401,8 @@ options! {
         "keep hygiene data after analysis (default: no)"),
     layout_seed: Option<u64> = (None, parse_opt_number, [TRACKED],
         "seed layout randomization"),
+    link_directives: bool = (true, parse_bool, [TRACKED],
+        "honour #[link] directives in crates (default: yes)"),
```
Comment (petrochenkov) [inline compiler/rustc_session/src/options.rs:1405]:
```suggestion
        "honor #[link] directives in the compiled crate (default: yes)"),
```
Need to emphasize that it's about *this* crate and not all crates like `-Zlink_native_libraries`.

(Also American English.)

Comment (jsgf) [reply compiler/rustc_session/src/options.rs]: OK, I pushed an update.

**Comment by jsgf** [comment]
I've deployed this internally and it's already showing value - when importing third-party packages, we no longer need to edit the source to remove `#[link]` directives.

**Comment by pcwalton** [comment]
Seems fine to me. Perhaps we should proactively remove the other `Zlink-native-libraries` flag as a "failed experiment"?

**Comment by jsgf** [comment]
> Perhaps we should proactively remove the other Zlink-native-libraries flag as a "failed experiment"?

I'll put up a separate PR in case someone is still using it.

**Comment by jsgf** [comment]
Ping @wesleywiser. Or @joshtriplett is this something you could accept?

**Comment by joshtriplett** [comment]
:+1: from me but I don't think I'm authoritative on compiler CLI arguments; someone from @rust-lang/compiler should r+ this.

**Review by davidtwco — APPROVED** [review]
r=me after rebase, this looks good to me and it's an unstable flag so we don't need to pin down the final name or take a vote or anything.

**Comment by jsgf** [comment]
@davidtwco OK, I rebased and fixed the conflicts.

**Comment by jsgf** [comment]
@pcwalton 
> Seems fine to me. Perhaps we should proactively remove the other Zlink-native-libraries flag as a "failed experiment"?

I think we still need `-Zlink-native-libraries` to deal with prebuilt rlibs - ie, the sysroot libraries.

**Pull request merged**